README.md
# Edustrate: the new way to learn.
### Open-Source code on GitHub

The files contained in this GitHub repository have been created, shared and uploaded by HD Software LTD, based in the UK. We reserve the right to refuse any pull requests or revoke a downloaded package/release.

This program has been designed and built by:
* Isaac Whitaker
* Oliver Whale
* Viggo Bryant-Frost
* Aisling Strugnell
* Bethan Cooling
* Theo Fletcher

You are permitted, under our Terms of Use, to use this code to develop applications that can work with Edustrate or addons for our marketplace (coming soon), but **you are not permitted, under any circumstance, to download this program and use it for your business, whether educational or not, without purchasing an official license from us.**

We do not at this time have a domain, but are in the process of purchasing one for the purposes of allowing you to be able to purchase a license, but you can contact us at **hdsoftwarehants@gmail.com** in the meantime.

### Versions Released
The most recent alpha version of Edustrate that has been released is v1.1.2-alpha, which adds in the Teacher UI feature to the program. Edustrate is expected to be officially released by v2.0, due by December 2021.

### Contributions
This code is, of course, open source, however there are some purposes we do not wish Edustrate to be used for, such as those listed above in Paragraph 3. We welcome contributions and ideas from third-party contributors and internal contributors, but HODs or Development need to approve the pull request before it can be merged with the `master` branch. (You will need to create a new branch when you commit something. It might ask you whether it wants you to commit directly to `master`: do not! Create a new one, and call it something like `username-patch-1`. It will be deleted if it is approved, so please don't expect it to last forever!).

&COPY; 2020 by HD Software, an imprint of Ablaze LLC. We do not own the right to the template on which the web app is built, so we have accredited them in our v1.1 release. All other rights reserved.


CHANGELOG.md
# Version History: Edustrate
We are documenting our web application updates and version history here. You can access the constantly updated version on GitHub.

## Our semantic system
* All of the first digits are distinctly major updates, for example a massive relocation and/or redesign.
* All of the second digits are less major updates, for example a new feature added to the existing app.
* All of the third digits are minor updates, for example a large bug fix or a new minor feature.
* All of the fourth digits are bug fixes or pre-release versions.

***

## Alpha Versions
These are pre-official release versions of Edustrate, up to date as of 9pm on May 26th 2020.

### v1
* Designed wireframed web application, showing index, task dashboard and other pages and their links in the sidebar
* Edited dropdowns and modified title bar

### v1.1
* Added official GitHub release
* Added about.html webpage listing details about the program and its design

### v1.1.0.1
* Added more HTML webpage (chatdash.html) which is the starting point for the Chat function.

### v1.1.2
* Created Teacher UI
* Edited structure: main folder is called `src`, subfolders are `.child`, `.teacher` and `icons`

**As of 26th May 2020, no other versions have been released. You can track updates via our GitHub repository.**


## Release Dates
| Version number | Release date        | Released on GitHub|
|-----------------|------------------|--------------------|
| < v1.0 | Multiple dates | ❌ |
| v1.0                     | April 2020 (date unknown) | ❌ |
| v1.1 | 26th May 2020 | ✔️ |
| v1.1.0.1 | 27th May 2020 | ✔️ |
| v1.1.2 | 28th May 2020 | ✔️ |


## Planned Folder Structure

Files are stored in a `.zip` file, with the syntax as follows:

```
edustrate [type: alpha, beta, stable, pre-release] v[number: 1.1.1.1].zip
```

<ul>
<li><code>src</code></li>
<li>README.md</li>
<li>CHANGELOG.md</li>
<li>LICENSE</li>
<ul>
<li><code>.child</code></li>
<li><code>.teacher</code></li>
<li><code>.parent</code></li>
<li><code>.tp</code></li>
<li><code>.pitstop</code></li>
<li><code>.globalassets</code></li>
</ul>
</ul>


### Bugs to be fixed & features to be added in `v1.1.3`
* A bug has been identified in the `.zip` file attached in the release v1.1.2 on GitHub. It means that the `.teacher` partition does not exist, and the UI cannot be viewed. Due to a storage fault, we can't now access the  original UI, but we are redesigning it for the occasion.
* We are now creating a parent UI in the zip file, in a new `.parent`partition. This will also feature a fix which allows you to view the child's dashboard.


v1.1.2/src/.teacher/startbootstrap-sb-admin-2-gh-pages/js/demo/chart-pie-demo.js
// Set new default font family and font color to mimic Bootstrap's default styling
Chart.defaults.global.defaultFontFamily = 'Nunito', '-apple-system,system-ui,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,sans-serif';
Chart.defaults.global.defaultFontColor = '#858796';

// Pie Chart Example
var ctx = document.getElementById("myPieChart");
var myPieChart = new Chart(ctx, {
  type: 'doughnut',
  data: {
    labels: ["On time", "Late", "Absent", "Sick"],
    datasets: [{
      data: [45, 9, 5, 1],
      backgroundColor: ['#1cc88a', '#f6c23e', '#e74a3b', '#4e73df'],
      hoverBackgroundColor: ['#008000', '#e68a00', '#b30000'],
      hoverBorderColor: "rgba(234, 236, 244, 1)",
    }],
  },
  options: {
    maintainAspectRatio: false,
    tooltips: {
      backgroundColor: "rgb(255,255,255)",
      bodyFontColor: "#858796",
      borderColor: '#dddfeb',
      borderWidth: 1,
      xPadding: 15,
      yPadding: 15,
      displayColors: false,
      caretPadding: 10,
    },
    legend: {
      display: false
    },
    cutoutPercentage: 80,
  },
});


v1.1.2/src/.teacher/startbootstrap-sb-admin-2-gh-pages/js/demo/chart-area-demo.js
// Set new default font family and font color to mimic Bootstrap's default styling
Chart.defaults.global.defaultFontFamily = 'Nunito', '-apple-system,system-ui,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,sans-serif';
Chart.defaults.global.defaultFontColor = '#858796';

function number_format(number, decimals, dec_point, thousands_sep) {
  // *     example: number_format(1234.56, 2, ',', ' ');
  // *     return: '1 234,56'
  number = (number + '').replace(',', '').replace(' ', '');
  var n = !isFinite(+number) ? 0 : +number,
    prec = !isFinite(+decimals) ? 0 : Math.abs(decimals),
    sep = (typeof thousands_sep === 'undefined') ? ',' : thousands_sep,
    dec = (typeof dec_point === 'undefined') ? '.' : dec_point,
    s = '',
    toFixedFix = function(n, prec) {
      var k = Math.pow(10, prec);
      return '' + Math.round(n * k) / k;
    };
  // Fix for IE parseFloat(0.55).toFixed(0) = 0;
  s = (prec ? toFixedFix(n, prec) : '' + Math.round(n)).split('.');
  if (s[0].length > 3) {
    s[0] = s[0].replace(/\B(?=(?:\d{3})+(?!\d))/g, sep);
  }
  if ((s[1] || '').length < prec) {
    s[1] = s[1] || '';
    s[1] += new Array(prec - s[1].length + 1).join('0');
  }
  return s.join(dec);
}

// Area Chart Example
var ctx = document.getElementById("myAreaChart");
var myLineChart = new Chart(ctx, {
  type: 'line',
  data: {
    labels: ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"],
    datasets: [{
      label: "Credits awarded",
      lineTension: 0.3,
      backgroundColor: "rgba(78, 115, 223, 0.05)",
      borderColor: "#e74a3b",
      pointRadius: 3,
      pointBackgroundColor: "#e74a3b",
      pointBorderColor: "#e74a3b",
      pointHoverRadius: 5,
      pointHoverBackgroundColor: "#e74a3b",
      pointHoverBorderColor: "#e74a3b",
      pointHitRadius: 10,
      pointBorderWidth: 2,
      data: [7, 3, 1, 5, 2],
    }],
  },
  options: {
    maintainAspectRatio: false,
    layout: {
      padding: {
        left: 10,
        right: 25,
        top: 25,
        bottom: 0
      }
    },
    scales: {
      xAxes: [{
        time: {
          unit: 'date'
        },
        gridLines: {
          display: false,
          drawBorder: false
        },
        ticks: {
          maxTicksLimit: 7
        }
      }],
      yAxes: [{
        ticks: {
          maxTicksLimit: 5,
          padding: 10,
          // Include a dollar sign in the ticks
          callback: function(value, index, values) {
            return number_format(value);
          }
        },
        gridLines: {
          color: "rgb(234, 236, 244)",
          zeroLineColor: "rgb(234, 236, 244)",
          drawBorder: false,
          borderDash: [2],
          zeroLineBorderDash: [2]
        }
      }],
    },
    legend: {
      display: false
    },
    tooltips: {
      backgroundColor: "rgb(255,255,255)",
      bodyFontColor: "#858796",
      titleMarginBottom: 10,
      titleFontColor: '#6e707e',
      titleFontSize: 14,
      borderColor: '#dddfeb',
      borderWidth: 1,
      xPadding: 15,
      yPadding: 15,
      displayColors: false,
      intersect: false,
      mode: 'index',
      caretPadding: 10,
      callbacks: {
        label: function(tooltipItem, chart) {
          var datasetLabel = chart.datasets[tooltipItem.datasetIndex].label || '';
          return datasetLabel + ": " + number_format(tooltipItem.yLabel);
        }
      }
    }
  }
});
